File: Data Structures & Algorithms/course-schedule/submission-3.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        preMap = defaultdict(list)
        for crs, pre in prerequisites:
            preMap[crs].append(pre)
        visit = set()

        def dfs(c):
            if c in visit:
                return False
            if preMap[c] == []:
                return True
            visit.add(c)
            for pre in preMap[c]:
                if not dfs(pre):
                    return False
            visit.remove(c)
            preMap[c] = []
            return True

        for c in range(numCourses):
            if not dfs(c):
                return False
        return True

File: Data Structures & Algorithms/counting-bits/submission-3.java
class Solution {
    public int[] countBits(int n) {
        int[] dp = new int[n + 1];
        for(int i = 1; i <= n; ++i) {
            dp[i] = dp[i >> 1] + (i & 1);
        }
        return dp;
    }
}

// 0 - 0000
// 1 - 0001
// 2 - 0010
// 3 - 0011
// 4 - 0100
// 5 - 0101
// 6 - 0110